### PR TITLE
Introduce resources to project API

### LINKED ISSUE #18881: Cannot access files packed inside the resources folder
**Description:**
When the .balo of a module is created, the resource files placed inside the resources folder is not accessible when executing the jar.

**Affected Versions:**
Ballerina 1.0.0 RC1

### LINKED ISSUE #22767: Add imported resources to the module jar
**Description:**
Currently, the module jar does not contain the imported module's resource within it.

Adding imported resources to the module jar is necessary.

### LINKED ISSUE #33652: Model module resources and test resources in the Project API
In the initial version of the project API design, I didn't model resources and test resources of a module. The API doesn't provide a way to access such resources. We've been getting several requests for this feature. I came across this while implementing the code generator support. https://github.com/ballerina-platform/ballerina-lang/issues/33651

### PR DESCRIPTION
**Pull request by azinneera**
Fixes #33652
Fixes #22767
Fixes #18881

## Approach

**Usage:**
1. Retrieve resources from a module
    ```java
    for (DocumentId documentId : module.resourceIds()) {
            Resource resource = module.resource(documentId);
            // do something with the resource
    }
    ```
   
2. Add a resource to a module
   ```java
   DocumentId documentId = DocumentId.create("log4j2.xml", defaultModule.moduleId());
   ResourceConfig resourceConfig = ResourceConfig.from(documentId, "log4j2.xml", contentStr.getBytes());
   module.modify().addResource(resourceConfig).apply();
   ```
3. Remove a resource from a module
    ```java
   module.modify().removeResource(documentId);
   ```

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/io/ballerina/projects/DocumentKind.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+ *  specific language governing permissions and limitations
+ *  under the License.
+ */
+package io.ballerina.projects;
+
+/**
+ * Represents the kind of the {@code Document} instance.
+ * <p>
+ * There are two known kinds of documents: USER_PROVIDED and GENERATED.
+ *
```
Comment (sameerajayasoma) [inline compiler/ballerina-lang/src/main/java/io/ballerina/projects/DocumentKind.java:24]: Do we need to make this a public API for now?

Comment (azinneera) [reply compiler/ballerina-lang/src/main/java/io/ballerina/projects/DocumentKind.java:24]: No not required to be exposed. We can keep it private or not include it at all

Comment (azinneera) [reply compiler/ballerina-lang/src/main/java/io/ballerina/projects/DocumentKind.java]: Removed the DocumentKind in ab4ba20

File: compiler/ballerina-lang/src/main/java/io/ballerina/projects/Module.java [hunk 2]
```diff
@@ -181,6 +210,46 @@ Modifier updateDocument(DocumentContext newDocContext) {
...
+            return this;
+        }
+
+        /**
+         * Creates a copy of the existing module and removes the specified resource from the new module.
+         *
+         * @param documentId documentId of the resource to remove
+         * @return an instance of the Module.Modifier
+         */
+        public Modifier removeResource(DocumentId documentId) {
```
Comment (sameerajayasoma) [inline compiler/ballerina-lang/src/main/java/io/ballerina/projects/Module.java:243]: Is it better to split this method into two? `removeResource` and `removeTestResource`

Comment (azinneera) [reply compiler/ballerina-lang/src/main/java/io/ballerina/projects/Module.java:243]: For bal documents, we have only `removeDocument` for both source files and test files. So followed the same approach for resources. Should we modify this only or removeDocument for the sake of consistency?

Comment (sameerajayasoma) [reply compiler/ballerina-lang/src/main/java/io/ballerina/projects/Module.java:243]: Let's keep `removeDocument` then.

**Review by sameerajayasoma — APPROVED** [review]

**Pull request merged**